File: backend/python-api/app/service/password_service.py
from __future__ import annotations

import asyncio
import logging

from passlib.hash import bcrypt

from app.domain.errors import ERR_INVALID_INPUT, ERR_TOKEN_INVALID, DomainError
from app.domain.password import validate_password_complexity
from app.domain.token import PasswordResetToken, generate_secure_token, token_expiry
from app.platform.email import EmailSender
from app.repository.token_repo import TokenRepository
from app.repository.user_repo import UserRepository
from app.service.token_service import TokenService

logger = logging.getLogger(__name__)


class PasswordService:
    def __init__(
        self,
        users: UserRepository,
        tokens: TokenRepository,
        token_svc: TokenService,
        email_sender: EmailSender,
        frontend_url: str,
    ):
        self._users = users
        self._tokens = tokens
        self._token_svc = token_svc
        self._email = email_sender
        self._frontend_url = frontend_url

    async def forgot_password(self, email: str, lang: str, session) -> None:
        try:
            user = await self._users.find_by_email(email)
        except DomainError:
            return

        await self._tokens.invalidate_password_reset_tokens(user.id)

        token_str = generate_secure_token()
        prt = PasswordResetToken(
            user_id=user.id,
            token=token_str,
            expires_at=token_expiry(15),
        )
        await self._tokens.create_password_reset_token(prt)
        await session.commit()

        link = f"{self._frontend_url}/reset-password?token={token_str}"
        asyncio.create_task(self._email.send_password_reset(user.email, link, lang))

    async def reset_password(self, token_str: str, new_password: str, session) -> None:
        prt = await self._tokens.find_password_reset_token(token_str)
        if prt is None:
            raise DomainError(ERR_TOKEN_INVALID, "Invalid or expired reset token")

        err = validate_password_complexity(new_password)
        if err:
            raise DomainError(ERR_INVALID_INPUT, err)

        user = await self._users.find_by_id(prt.user_id)
        user.password_hash = bcrypt.using(rounds=12).hash(new_password)
        await self._users.update(user)
        await self._tokens.mark_password_reset_token_used(prt.id)
        await self._token_svc.revoke_all_user_tokens(user.id)
        await session.commit()

    async def verify_email(self, token_str: str, session) -> None:
        evt = await self._tokens.find_email_verification_token(token_str)
        if evt is None:
            raise DomainError(ERR_TOKEN_INVALID, "Invalid or expired verification token")

        user = await self._users.find_by_id(evt.user_id)
        user.verified = True
        await self._users.update(user)
        await self._tokens.mark_email_verification_token_used(evt.id)
        await session.commit()

File: backend/go-api/internal/service/password_service.go
package service

import (
	"context"
	"errors"
	"log"
	"time"

	"saas-starter/backend/go-api/internal/domain"
	"saas-starter/backend/go-api/internal/platform/email"
	"saas-starter/backend/go-api/internal/repository"

	"golang.org/x/crypto/bcrypt"
)

type PasswordService struct {
	users       repository.UserRepository
	tokens      repository.TokenRepository
	txManager   repository.TxManager
	emailSender email.Sender
	frontendURL string
}

func NewPasswordService(
	users repository.UserRepository,
	tokens repository.TokenRepository,
	txManager repository.TxManager,
	emailSender email.Sender,
	frontendURL string,
) *PasswordService {
	return &PasswordService{
		users: users, tokens: tokens, txManager: txManager,
		emailSender: emailSender, frontendURL: frontendURL,
	}
}

func (s *PasswordService) ForgotPassword(ctx context.Context, emailAddr string) error {
	user, err := s.users.FindByEmail(ctx, emailAddr)
	if err != nil {
		if errors.Is(err, domain.ErrNotFound) {
			return nil // don't leak whether user exists
		}
		return err
	}

	s.tokens.InvalidatePasswordResetTokens(ctx, user.ID)

	tokenStr := domain.GenerateSecureToken()
	resetToken := &domain.PasswordResetToken{
		UserID:    user.ID,
		Token:     tokenStr,
		ExpiresAt: domain.TokenExpiry(15),
	}
	if err := s.tokens.CreatePasswordResetToken(ctx, resetToken); err != nil {
		return err
	}

	link := s.frontendURL + "/reset-password?token=" + tokenStr
	lang := user.PreferredLang("en")

	go func() {
		if err := s.emailSender.SendPasswordReset(user.Email, link, lang); err != nil {
			log.Printf("Failed to send reset email to %s: %v", user.Email, err)
		}
	}()

	return nil
}

func (s *PasswordService) ResetPassword(ctx context.Context, tokenStr, newPassword string) error {
	resetToken, err := s.tokens.FindPasswordResetToken(ctx, tokenStr)
	if err != nil {
		return domain.NewError(domain.ErrTokenInvalid, "invalid or expired token")
	}

	if time.Now().After(resetToken.ExpiresAt) {
		return domain.NewError(domain.ErrTokenExpired, "token expired")
	}

	if err := domain.ValidatePasswordComplexity(newPassword); err != nil {
		return domain.NewError(domain.ErrInvalidInput, err.Error())
	}

	user, err := s.users.FindByID(ctx, resetToken.UserID)
	if err != nil {
		return err
	}

	hash, err := bcrypt.GenerateFromPassword([]byte(newPassword), BcryptCost)
	if err != nil {
		return err
	}
	user.PasswordHash = string(hash)

	return s.txManager.WithTx(ctx, func(txCtx context.Context) error {
		if err := s.users.Update(txCtx, user); err != nil {
			return err
		}
		if err := s.tokens.MarkPasswordResetTokenUsed(txCtx, resetToken.ID); err != nil {
			return err
		}
		return s.tokens.RevokeAllUserRefreshTokens(txCtx, user.ID)
	})
}

func (s *PasswordService) VerifyEmail(ctx context.Context, tokenStr string) error {
	vToken, err := s.tokens.FindEmailVerificationToken(ctx, tokenStr)
	if err != nil {
		return domain.NewError(domain.ErrTokenInvalid, "invalid or expired token")
	}

	if time.Now().After(vToken.ExpiresAt) {
		return domain.NewError(domain.ErrTokenExpired, "token expired")
	}

	user, err := s.users.FindByID(ctx, vToken.UserID)
	if err != nil {
		return err
	}

	user.Verified = true

	return s.txManager.WithTx(ctx, func(txCtx context.Context) error {
		if err := s.users.Update(txCtx, user); err != nil {
			return err
		}
		return s.tokens.MarkEmailVerificationTokenUsed(txCtx, vToken.ID)
	})
}
